### ISSUE TITLE
property-value-whitelist/blacklist

### ISSUE THREAD

**Issue by babichss** [stylelint/stylelint#551]
Add property-value-whitelist option to check allowed values for properties.

e.g. box-sizing: 1; should be reported as error if property-value-whitelist for this property is set to ["border-box", "padding-box"]

**Comment by jeddy3** [comment]
That's an interesting suggestion.

Are there many more use cases you can think of, besides `boxing-sizing`, where you'd want to restrict a property value to a _single_ keyword?

The [values spec](http://www.w3.org/TR/css3-values/#keywords) has the `collapse | separate` keyword values for the `border-collapse` property as an example.

Any others?

**Comment by Sergiy-Babich** [comment]
This should be useful for all properties with keywords as values, e.g. display, background-size, visibility, etc.

**Comment by jeddy3** [comment]
> display, background-size, visibility

Thanks! I was struggling to think of more off-the-top-of-my-head.

This looks like a good addition to me. @davidtheclark What do you think?

It also looks to be just a simplified version of `property-unit-whitelist`, where we're checking the whole value rather than parsing the value and checking each of the units.

We'll need to add `property-value-blacklist` for parity with the other blacklist/whitelist rules.

**Comment by evilebottnawi** [comment]
@Sergiy-Babich @jeddy3 Properties that contain error specification should not be checked. To do this, there are modules postcss. But `property-value-whitelist` and `property-value-blacklist` good idea. Example, disable use `all` for `transition`, because it is not the best use and reduces performance.

**Comment by davidtheclark** [comment]
How will we distinguish between the keywords that can be discriminated and other values that cannot be? For example, `background-size` can have keywords (`cover`, `contain`, `auto`, etc) but can also have measurements (`50%`, `6px`, etc). If you whitelist `cover` and `contain`, what do we do about `50%`? Or for `transition`, which @evilebottnawi mentioned, if you have the declaration `transition: all 1s linear;`, how do we know to only complain about `all`, not about `1s`?

**Comment by evilebottnawi** [comment]
@davidtheclark maybe add more powerfull `postcss-value-parser` for this goals. I think need discussion. About `transition: all 1s linear;`, scss-lint have rule for this, but solution is not general, as proposed here.

**Comment by davidtheclark** [comment]
Well, we could use the value parser and then assume that any `word` node starting with a number is not a keyword and can be ignored ... (does that sound right @TrySound?)

If somebody wants to develop this, we'd just need to make sure to have a lot of tests of a lot of different kinds of value syntaxes, with keywords, user-determined identifiers (like animation-name), functions, numbers, etc.

**Comment by TrySound** [comment]
@davidtheclark Seems good and simple. And I'm not sure new parser can give more information. It will be the same stuff, but with keyword detecting. Why not detect keywords separatly?

**Comment by evilebottnawi** [comment]
@davidtheclark about `transition` and not, properties may shorthand andseparate
```css
transition: all 2s linear 1s;
```
or
```css
transition-property: width;
    transition-duration: 2s;
    transition-timing-function: linear;
    transition-delay: 1s;
```
It is also worth considering.

**Comment by kangax** [comment]
I came looking for this as well. 

Definitely very useful and I'm surprised to see both property-blacklist and property-unit-blacklist but not property-value-blacklist.

In our case, "text-transform: uppercase" is not i18n friendly so we'd like to warn against that. Then there's potential performance problems with "position: fixed" or "transform: 3d..." on mobile. Also opacity that's >0 && <1 is a potential performance issue on any platform.

I can contribute this rule.

The simplest API would be (by convention with others) `{ "property": ["array", "of", "values"] }` but in our case I'm afraid we'll need regex too at the very least.

1) `text-transform: uppercase` and `position: fixed` are easy:

```js
{
  "text-transform": ["uppercase"],
  "position": ["fixed"]
}
```

2) transform: 3d... is harder. Either we list all 3d properties (which is fragile and not future-proof) or we need support for regex and/or function:

```js
{
  "transform": [/3d\(/],
  // or (least flexible)
  "transform": ["scale3d", "rotate3d", "translate3d", "matrix3d"],
  // or (most flexible)
  "transform": function(val) { return val.indexOf('3d(') > -1; }
}
```

3) opacity <1&&>0 (non-opaque) is best implemented with function as well:

```js
{
  "opacity": function(val) { return val > 0 && val < 1; }
}
```

Functions are problematic in configs and JSON though :/

Finally, last piece of the puzzle is custom messages. Our current linter shows messages such as:

```js
{
  title: "Poor performance alert: opacity property",
  explanation: "opacity is known to cause performance issues in browsers. Please be intentional and ..."
}
```

I'm currently porting it to PostCSS + stylelint and adding custom plugins that simply `walkDecls` and check `node.prop`, `node.value` for certain combination and report these messages and explanations.

It's the most flexible approach but also verbose.

One solution could be to do something like this:

```
{
  "position": [["fixed"], "position: fixed is very buggy on mobile. Please read ..."],
  "transform": [[/3d\(/], "These cause hardware acceleration in webkit. Massive performance ..."]
}
```

It could be useful to adapt this to other rules too. On the other hand, it could be a little too specific, so I would understand if you were against adding this complexity.

(I can still contribute simple property-value-blacklist and also add a custom plugin for our usecase)

**Comment by kangax** [comment]
Alternative formats:

```js
property-value-blacklist: {
  // either specify object, in which case we expect "values" and "message"
  "position": { 
    "values": [ "fixed" ], 
    "message": "..."
  },
  // or array, in which case there's no message
  "transform": [/3d\(/]
}
```

or slightly different take:

```js
property-value-blacklist: [
  { "property": "position", "values": [...], "message": "..." },
  { "property": "transform", "values": [...], "message": "..."}
]
```

**Comment by davidtheclark** [comment]
@kangax Thanks for pitching in.

If we are going to allow for custom messages I think it would be best if we could come up with a solution that would work generally, not just for this rule. (Though I do understand why custom messages would be especially useful for this particular rule.) An issue exists for this: https://github.com/stylelint/stylelint/issues/400 --- maybe we can talk about ideas over there?

I'm afraid that functions are probably not an option for rule configuration, because they would not be available with JSON and YAML config files or `package.json` config properties. 

The `opacity` issue could be a rule of its own: if it has a good justification, people may want to use it without having to set up `property-value-blacklist`. I'm curious: do you have any things to read about this? I hadn't heard of a problem before, and it seems that if you rule out every number between 0 and 1 you've pretty much ruled out the use of opacity, right (i.e. what not use visibility)? Or am I misunderstanding?

I think the idea of allowing regular expressions would add a lot of power to this rule, and I can't immediately think of any clear downside. They would just have to be strings instead of actual JS constructs. We could make it a convention that if the string starts and ends with `/` we feed what's between the strings into a regular expression, e.g. `"/3d\\(/"`???

**Comment by kangax** [comment]
@davidtheclark Sure, I'll read through #400. Not being able to add functions — understandable. Regexes could definitely solve a lot. As far as opacity, you're right that not using 1 or 0 is almost like not using opacity at all. This lint rule is really just an advice. People will still use opacity; we just want them to be mindful about it and understand that it could cause performance issues (http://cubiq.org/performance-tricks-for-mobile-web-development).

I guess there's few things to do here:

1) Implement property-value-blacklist as usual (like all other existing similar rules)
2) Discuss and add support for regexes as a separate issue
3) Discuss and add support for custom messages in #400 

Does that sound reasonable?

**Comment by davidtheclark** [comment]
Great! I think we should add regexes as part of the first implementation of property-value-blacklist (it seems like a good idea to me) --- so 1 and 2 can go together, then 3 stays out as a separate issue. 

(If we decide to do use regexes in other rules, then, we'll just genericize the technique in this rule and re-use it.)

Does that make sense?

**Comment by kangax** [comment]
Sounds good, I'll get on it.

**Comment by jeddy3** [comment]
@davidtheclark @kangax Are we intending to build a companion `property-value-whitelist` rule? Or shall we close this issue? We can always open a new issue for `property-value-whitelist` when/if someone voices a need for it.

**Comment by kangax** [comment]
@jeddy3 definitely need it for consistency but I've got some other things to take care of now (and we also only need blacklist) so would be great if someone else could add!

**Comment by jeddy3** [comment]
@kangax FYI, `4.3` is out with `property-value-blacklist`. Along with `property-value-whitelist` and a couple of the things you worked-on/requested e.g. exposing the rules.

Thanks again for knocking out `property-value-blacklist`! :)

**Comment by kangax** [comment]
Awesome, thanks!

**Comment by ZacharyRSmith** [comment]
Was `property-value-blacklist` removed?  I was hoping to use it or a `value-blacklist` to warn against IE-unsupported values like `unset` and `initial`.  Thx!

**Comment by ZacharyRSmith** [comment]
Nevermind, I see `declaration-property-value-blacklist` is probably what `property-value-blacklist` became.

**Comment by jeddy3** [comment]
@ZacharyRSmith That's right, it became `declaration-property-value-blacklist` to bring it in line with our [standardised](https://github.com/stylelint/stylelint/blob/master/docs/user-guide/rules.md) [names](https://github.com/stylelint/stylelint/blob/master/docs/user-guide/about-rules.md).

**Comment by hudochenkov** [comment]
@theronpatrick correct config:

```json
{
  "rules": {
    "declaration-property-value-blacklist": [
      {
        "text-transform": ["uppercase"]
      },
      { "message": "CUSTOM MESSAGE" }
    ]
  }
}
```

You can specify custom message for the whole rule only.
